### PR TITLE
Fix race condition in 'test_compat'

### PR DESCRIPTION
**Pull request by oleorhagen**
Changelog: None
Signed-off-by: Ole Petter <ole.orhagen@northern.tech>

### CODE REVIEW

File: tests/tests/test_compat.py [hunk 1]
```diff
@@ -189,40 +189,68 @@ def assert_successful_deployment(api_deployments, deployment_id, timeout=TIMEOUT
...
+                RebootDetectorThread(host_ip, client, timeout, self.enter_barrier)
+            )
+
+    def __enter__(self):
+        for d in self.reboot_detectors:
+            d.start()
+
+        # Wait for all reboot-detectors to finish their setup, before starting
+        # the code within the 'with' statement
+        self.enter_barrier.wait()
```
Comment (alfrunes) [inline tests/tests/test_compat.py:247]: Is it 100% certain that the parallel threads can't raise an exception? Can you add a short timeout here (~10sec) just to be safe? And also abort the test `if self.enter_barrier.broken` after.

**Review by kacf — APPROVED** [review]
Very nice solution!

**Review by alfrunes — CHANGES_REQUESTED** [review]
Well done!
Maybe we should consider replacing (or merging the parallel implementation into) the reboot detector in testutils as well?
I just have one small change request.

File: tests/tests/test_compat.py (hunk 1 above)
Comment (oleorhagen) [reply tests/tests/test_compat.py:247]: As far as I understand, all exceptions are caught and put into the exception queue? And can the barrier be broken if no one has called `abort`?

**Comment by oleorhagen** [comment]
> Well done!
> Maybe we should consider replacing (or merging the parallel implementation into) the reboot detector in testutils as well?
> I just have one small change request.

Yes I was playing with the idea. Need some more work then though, as this one only supports one reboot. How about we merge this, and I promise I will follow up!

File: tests/tests/test_compat.py (hunk 1 above)
Comment (alfrunes) [reply tests/tests/test_compat.py:247]: I believe the barrier "breaks" if you time out waiting for it.

File: tests/tests/test_compat.py [hunk 2]
```diff
@@ -189,40 +189,68 @@ def assert_successful_deployment(api_deployments, deployment_id, timeout=TIMEOUT
...
+        self._timeout = timeout
+        self._exception_q = queue.Queue(maxsize=1)
+        self.detector = self._client.get_reboot_detector(self._host_ip)
+        self.barrier = barrier
+
+    def run(self):
+        try:
+            # Setup the individual reboot-detectors
+            with self.detector as rd:
+                # Possibly move this to after the reboot-detector accepts connections (?)
```
Comment (alfrunes) [inline tests/tests/test_compat.py:206]: ... I say yes, and then time out if the client does not respond (ref. review comment).

File: tests/tests/test_compat.py (hunk 1 above)
Comment (oleorhagen) [reply tests/tests/test_compat.py:247]: > This method may raise a BrokenBarrierError exception if the barrier is broken or reset while a thread is waiting.

Is this not good enough? I mean, the test will fail with a broken barrier exception

File: tests/tests/test_compat.py (hunk 2 above)
Comment (oleorhagen) [reply tests/tests/test_compat.py:206]: I think I'm going to leave it as it is actually. This is easier, and cleaner. It fixes the recurring issue we have been seeing. The comment is a little misleading now, as it was made before some of the changes.

**Review by alfrunes — APPROVED** [review]

**Pull request merged**